lighthouse-core/lib/i18n.js

```diff
@@ -43,57 +41,150 @@ const formats = {
+      if (pathAsString.length) pathAsString += '.';
+      pathAsString += property;
+    } else {
+      if (/]|"|'/.test(property)) throw new Error(`Cannot handle "${property}" in i18n`);
+      pathAsString += `[${property}]`;
+    }
+  }
+
+  return pathAsString;
 }
 
 module.exports = {
   UIStrings,
+  formatPathAsString,
+  /**
+   * @return {LH.Locale}
+   */
+  getDefaultLocale() {
+    const defaultLocale = MessageFormat.defaultLocale;
+    if (defaultLocale in LOCALES) return /** @type {LH.Locale} */ (defaultLocale);
+    return 'en-US';
+  },
   /**
    * @param {string} filename
    * @param {Record<string, string>} fileStrings
    */
   createStringFormatter(filename, fileStrings) {
     const mergedStrings = {...UIStrings, ...fileStrings};
 
-    /** @param {string} msg @param {*} [values] */
-    const formatFn = (msg, values) => {
-      const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === msg);
-      if (!keyname) throw new Error(`Could not locate: ${msg}`);
-      preprocessMessageValues(msg, values);
+    /** @param {string} template @param {*} [values] */
+    const formatFn = (template, values) => {
+      const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === template);
+      if (!keyname) throw new Error(`Could not locate: ${template}`);
 
       const filenameToLookup = keyname in UIStrings ? __filename : filename;
-      const lookupKey = path.relative(LH_ROOT, filenameToLookup) + '!#' + keyname;
-      const localeStrings = LOCALES[locale] || {};
-      const localeString = localeStrings[lookupKey] && localeStrings[lookupKey].message;
-      // fallback to the original english message if we couldn't find a message in the specified locale
-      // better to have an english message than no message at all, in some number cases it won't even matter
-      const messageForMessageFormat = localeString || msg;
-      // when using accented english, force the use of a different locale for number formatting
-      const localeForMessageFormat = locale === 'en-XA' ? 'de-DE' : locale;
-
-      const formatter = new MessageFormat(messageForMessageFormat, localeForMessageFormat, formats);
-      return formatter.format(values);
+      const unixStyleFilename = path.relative(LH_ROOT, filenameToLookup).replace(/\\/g, '/');
+      const key = unixStyleFilename + '!#' + keyname;
+      const keyUsages = formattedStringUsages.get(key) || [];
+      keyUsages.push({key, template, values});
+      formattedStringUsages.set(key, keyUsages);
+
+      return `${key}#${keyUsages.length - 1}`;
     };
 
     return formatFn;
   },
   /**
-   * @param {LH.Locale|null} [newLocale]
+   * @param {LH.Result} lhr
+   * @param {LH.Locale} locale
    */
-  setLocale(newLocale) {
-    if (!newLocale) return;
-    locale = newLocale;
+  replaceLocaleStringReferences(lhr, locale) {
+    /**
+     * @param {*} objectInLHR
+     * @param {LH.LocaleLog} log
+     * @param {string[]} path
```

Comment:
only for testing?

Reply:
are you referring to why path is an array of strings in this function signature? it's so that we can easily push onto it and convert the name at each step rather than have two functions that build from an array *and* one that builds incrementally